File: Codeforces/1221/D.py
def calc(add, pos):

    if dp[add][pos] != -1:
       return dp[add][pos]

    if pos == n:
        dp[add][pos] = 0
        return 0

    dp[add][pos] = float('inf')

    for x in range(0,3):
        if pos == 0 or (a[pos]+x != a[pos-1] + add):
            dp[add][pos] = min(dp[add][pos], calc(x, pos+1) + x * b[pos])
            
    return dp[add][pos]


q = int(input())

dp = [[-1] * 300001 for l in range(3)]
a = [0] * 300001
b = [0] * 300001

for _ in range(q):
    n = int(input())
    
    for p in range(n):
        x,y = [int(i) for i in input().split()]
        a[p] = x
        b[p] = y
        dp[0][p] = -1
        dp[1][p] = -1
        dp[2][p] = -1

    print(calc(0,0))

File: Codeforces/1221/D.cpp
#include <bits/stdc++.h>

#define ll long long
#define MAX int(3e5+10)

using namespace std;


int n;
int a[MAX];
int b[MAX];
ll dp[MAX][3];

ll calc(int pos, int add){
    
    ll &res = dp[pos][add];
    
    if (res != -1)
        return res;
        
    res = (ll)1e18 + 100;

    if(pos == n) 
        return res = 0;
        
    for(ll i=0; i<3; i++){
        // if we're at pos 0 or 
        // if current pos + what we are adding now is different from last pos + what we added to it
        if(pos == 0 || a[pos] + i != a[pos - 1] + add)
            // we pass to the next state the next position and the added value to the current position
            res = min(res, calc(pos + 1, i) + i * b[pos]);
    }

    return res;
}


int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int t;
    cin >> t;
    
    while(t--){
        cin >> n;
        for(int i=0; i < n; i++){
            cin >> a[i];
            cin >> b[i];
        }

        for(int i=0; i <= n; i++){
            dp[i][0] = dp[i][1] = dp[i][2] = -1;
        }
        
        cout << calc(0,0) << "\n";
    }
    return 0;
}
